%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,12pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax

\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
 \ifdefined\DeclareUnicodeCharacterAsOptional\else
  \DeclareUnicodeCharacter{00A0}{\nobreakspace}
\fi\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage[dontkeepoldnames]{sphinx}

\usepackage{geometry}

% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}
\addto\captionsenglish{\renewcommand{\contentsname}{Contents:}}

\addto\captionsenglish{\renewcommand{\figurename}{Fig.}}
\addto\captionsenglish{\renewcommand{\tablename}{Table}}
\addto\captionsenglish{\renewcommand{\literalblockname}{Listing}}

\addto\extrasenglish{\def\pageautorefname{page}}

\setcounter{tocdepth}{1}



\title{dfoil Documentation}
\date{Nov 07, 2017}
\release{2017-08-25}
\author{James B. Pease}
\newcommand{\sphinxlogo}{\sphinxincludegraphics{logo.png}\par}
\renewcommand{\releasename}{Release}
\makeindex

\begin{document}

\maketitle
\sphinxtableofcontents
\phantomsection\label{\detokenize{index::doc}}



\chapter{Getting Started}
\label{\detokenize{intro:welcome-to-dfoil-s-documentation}}\label{\detokenize{intro:getting-started}}\label{\detokenize{intro::doc}}

\section{What is dfoil?}
\label{\detokenize{intro:what-is-dfoil}}
\sphinxstyleemphasis{D}$_{\text{FOIL}}$is a method for testing introgression in a five-taxon symmetric phylogeny.


\section{How do I cite this sofware?}
\label{\detokenize{intro:how-do-i-cite-this-sofware}}
If you use this program, please cite:

\sphinxcode{James B Pease, Matthew W. Hahn. 2015.
"Detection and Polarization of Introgression in a Five-taxon Phylogeny"
Systematic Biology. 64 (4): 651-662.
http://www.dx.doi.org/10.1093/sysbio/syv023
doi: 10.1093/sysbio/syv023}

Please also include the link \textless{}\sphinxurl{https://www.github.com/jbpease/dfoil}\textgreater{} in your publication.


\section{Requirements}
\label{\detokenize{intro:requirements}}\begin{itemize}
\item {} 
Python 2.7.x or 3.x

\item {} 
Scipy: \sphinxurl{http://www.scipy.org/}

\item {} 
Numpy: \sphinxurl{http://www.numpy.org/}

\item {} 
matplotlib 1.5.3+: \sphinxurl{http://www.matplotlib.org/}

\end{itemize}


\subsection{Optional}
\label{\detokenize{intro:optional}}
To run simulations within dfoil\_sim.py you will also need:
* ms: \sphinxurl{http://home.uchicago.edu/rhudson1/source/mksamples.htm}


\section{Installation}
\label{\detokenize{intro:installation}}
No installation is necessary, just download dfoil and run scripts through Python.
\sphinxcode{git clone https://www.github.com/jbpease/dfoil}
Other required software should be installed according to their individual instructions.


\section{Preparing your data}
\label{\detokenize{intro:preparing-your-data}}

\subsection{Generating an AB-pattern count file from a FASTA file.}
\label{\detokenize{intro:generating-an-ab-pattern-count-file-from-a-fasta-file}}
Use the script fasta2dfoil (see options below) to generate the count file from a FASTA.


\subsection{Using a prepared count file:}
\label{\detokenize{intro:using-a-prepared-count-file}}
One or more input files can be specified. These files can have any number of header lines at the beginning (including none), but \sphinxstylestrong{all header lines must start with ‘\#’}

Data fields should be tab/space separated files with two starting fields:
\begin{itemize}
\item {} 
\sphinxcode{CHROMOSOME} (this can be dummy values)

\item {} 
\sphinxcode{POSITION} (also can be dummy values)

\end{itemize}

followed by the pattern counts in this order:

\sphinxcode{AAAAA AAABA AABAA AABBA ABAAA ABABA ABBAA ABBBA, BAAAA BAABA BABAA BABBA BBAAA BBABA BBBAA BBBBA}

for the four-taxon test the patterns are
\sphinxcode{AAAA AABA ABAA ABBA BAAA BABA BBAA BBBA}

\begin{sphinxadmonition}{hint}{Hint:}
these patterns are in ‘binary’ order 0000, 0010, 0100…
\end{sphinxadmonition}

\begin{sphinxadmonition}{important}{Important:}
The order of taxa must be P1 P2 P3 P4 O, such that:
* “O” is the outgroup
* P1 and P2 are a monophyletic pair of taxa
* P3 and P4 are a monophyletic pair of taxa
* P3 and P4 divergence occurs before (in forward time) the divergence of P1 and P2
(The choice of P1/P2 and P3/P4 within the pairings is arbitrary)
\end{sphinxadmonition}


\section{Running dfoil.py}
\label{\detokenize{intro:running-dfoil-py}}

\subsection{Basic usage}
\label{\detokenize{intro:basic-usage}}
\sphinxcode{{}`python dfoil.py -{-}infile INPUTFILE1 {[}INPUTFILE2 ...{]} -{-}out OUTPUTFILE1 {[}OUTPUTFILE2 ...{]}{}`}


\subsection{Pre-check}
\label{\detokenize{intro:pre-check}}
The data will undergo a precheck (use \sphinxcode{-{-}skip-pre-check} to turn off, or \sphinxcode{-{-}pre-check-only} to only run the pre-check.
This will check for common issues in count data that might affect the result or violate the assumptions.

Common issues include:
* An accelerated rate of substitutions (i.e. an excess of B’s) on a specific branch relative to its sister taxon
* Mis-labeling of P1/P2 and P3/P4 (remember P3/P4 divergence should come first, in forward time)


\subsection{Modes}
\label{\detokenize{intro:modes}}\begin{itemize}
\item {} 
dfoil = standard DFOIL for five-taxa (this is default)

\item {} 
dfoilalt = dfoil without single-B patterns (ABAAA, etc.)

\item {} 
partitioned = Partitioned D-statistics (Eaton \& Ree 2013)

\item {} 
dstat = four-taxon D-statistic (Green et al. 2010)

\item {} 
dstatalt = four-taxon D-statistic with inverse patterns added (use with caution)

\end{itemize}


\subsection{Advanced Weight Parameters}
\label{\detokenize{intro:advanced-weight-parameters}}
The parameters \sphinxtitleref{\textendash{}beta1}, \sphinxtitleref{\textendash{}beta2}, and \sphinxtitleref{\textendash{}beta3} are weighting factors (0 \textless{}= b \textless{}= 1), for single-B, double-B, and triple-B patterns, respectively. Ordinarily you will not need to set these.

By default these are set as:
\begin{itemize}
\item {} 
dfoil = 1,1,1

\item {} 
dfoilalt = 0,1,1

\item {} 
dstat: 0,1,N/A (no triple-B)

\item {} 
dstatalt: 1,1,N/A (no triple-B)

\item {} 
partitioned: N/A (does not use weighting parameters)

\end{itemize}


\section{Output Format}
\label{\detokenize{intro:output-format}}
One or more output files are specified (equal to number of inputs).
The outputs will have fields:
\begin{itemize}
\item {} 
\sphinxcode{CHROMOSOME}

\item {} 
\sphinxcode{POSITION}

\end{itemize}

then for each D-statistic:
\begin{itemize}
\item {} 
\sphinxcode{Dxx\_left} (left term value)

\item {} 
\sphinxcode{Dxx\_right} (right term value)

\item {} 
\sphinxcode{Dxx\_stat} (D-statistic value)

\item {} 
\sphinxcode{Dxx\_chisq} (Chi\_Squared value)

\item {} 
\sphinxcode{Dxx\_pval} (Chi\_Squared P-value)

\end{itemize}

(where “xx” will be replaced with FO, IL, FI, OL)


\section{License}
\label{\detokenize{intro:license}}
This file is part of dfoil.

dfoil is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as publihed by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

dfoil is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with Foobar.  If not, see (\sphinxurl{http://www.gnu.org/licenses/}).


\chapter{Program Parameter Descriptions}
\label{\detokenize{prog_desc:program-parameter-descriptions}}\label{\detokenize{prog_desc::doc}}

\section{dfoil}
\label{\detokenize{prog_desc:dfoil}}

\subsection{Description}
\label{\detokenize{prog_desc:description}}
DFOIL: Directional introgression testing a five-taxon phylogeny
dfoil - Calculate DFOIL and D-statistics stats from one or more count files.
James B. Pease
\sphinxurl{http://www.github.com/jbpease/dfoil}

USAGE: dfoil.py INPUTFILE1 … \textendash{}out OUTPUTFILE1 …


\subsection{Parameters}
\label{\detokenize{prog_desc:parameters}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:h-help}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}infile} (required)}
\label{\detokenize{prog_desc:infile-required}}
\sphinxstylestrong{Description:} input tab-separated counts file

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}out} (required)}
\label{\detokenize{prog_desc:out-required}}
\sphinxstylestrong{Description:} outputs tab-separated DFOIL stats

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}beta1}}
\label{\detokenize{prog_desc:beta1}}\begin{description}
\item[{\sphinxstylestrong{Description:} beta1 coefficient for single-B patterns,}] \leavevmode
defaults: DFOIL/Dstatalt=1.0,
DFOILalt,Dstat=0,Dpart=N.A.

\end{description}

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}beta2}}
\label{\detokenize{prog_desc:beta2}}\begin{description}
\item[{\sphinxstylestrong{Description:} beta2 coefficient for double-B patterns,}] \leavevmode
defaults: Dpart=N.A., others=1.0

\end{description}

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} 1.0


\subsubsection{\sphinxstyleliteralintitle{-{-}beta3}}
\label{\detokenize{prog_desc:beta3}}\begin{description}
\item[{\sphinxstylestrong{Description:} beta3 coefficient for triple-B patterns}] \leavevmode
defaults: DFOIL/DFOILalt=1.0,
Dstat/Dpart=N.A.

\end{description}

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} 1.0


\subsubsection{\sphinxstyleliteralintitle{-{-}mincount}}
\label{\detokenize{prog_desc:mincount}}
\sphinxstylestrong{Description:} minium number of D denominator sites per window

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 10


\subsubsection{\sphinxstyleliteralintitle{-{-}mintotal}}
\label{\detokenize{prog_desc:mintotal}}
\sphinxstylestrong{Description:} minimum total number of sites in a region

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 50


\subsubsection{\sphinxstyleliteralintitle{-{-}mode}}
\label{\detokenize{prog_desc:mode}}\begin{description}
\item[{\sphinxstylestrong{Description:} dfoil = DFOIL,}] \leavevmode
dfoilalt = DFOIL without single-B patterns,
partitioned = Partitioned D-statistics,
dstat = Four-Taxon D-statistic,
dstatalt = Four-Taxon D-statistic
with single-B patterns

\end{description}

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} dfoil

\sphinxstylestrong{Choices:} {[}‘dfoil’, ‘dfoilalt’, ‘partitioned’, ‘dstat’, ‘dstatalt’{]}


\subsubsection{\sphinxstyleliteralintitle{-{-}plot}}
\label{\detokenize{prog_desc:plot}}
\sphinxstylestrong{Description:} write plot to file path(s) given

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}plot\_background}}
\label{\detokenize{prog_desc:plot-background}}\begin{description}
\item[{\sphinxstylestrong{Description:} 0-1.0 background intensity}] \leavevmode
(0=none, default=0.3)

\end{description}

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} 0.3


\subsubsection{\sphinxstyleliteralintitle{-{-}plot\_color}}
\label{\detokenize{prog_desc:plot-color}}
\sphinxstylestrong{Description:} choose color mode

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} color

\sphinxstylestrong{Choices:} {[}‘color’, ‘colordark’, ‘bw’, ‘bwdark’{]}


\subsubsection{\sphinxstyleliteralintitle{-{-}plot\_height}}
\label{\detokenize{prog_desc:plot-height}}
\sphinxstylestrong{Description:} height of plot (in cm)

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} 8.0


\subsubsection{\sphinxstyleliteralintitle{-{-}plot\_hideaxes}}
\label{\detokenize{prog_desc:plot-hideaxes}}
\sphinxstylestrong{Description:} hide axes labels

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}plot\_hidekey}}
\label{\detokenize{prog_desc:plot-hidekey}}
\sphinxstylestrong{Description:} hide plot key

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}plot\_labels}}
\label{\detokenize{prog_desc:plot-labels}}
\sphinxstylestrong{Description:} taxon labels

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}plot\_lineweight}}
\label{\detokenize{prog_desc:plot-lineweight}}
\sphinxstylestrong{Description:} line weight for dplots (default=1pt)

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} 1.0


\subsubsection{\sphinxstyleliteralintitle{-{-}plot\_noanc}}
\label{\detokenize{prog_desc:plot-noanc}}\begin{description}
\item[{\sphinxstylestrong{Description:} do not plot background for}] \leavevmode
ancestral introgression

\end{description}

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}plot\_smooth}}
\label{\detokenize{prog_desc:plot-smooth}}
\sphinxstylestrong{Description:} average D-stats over this number of points

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}plot\_totals}}
\label{\detokenize{prog_desc:plot-totals}}
\sphinxstylestrong{Description:} add a background plot of total site counts

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}plot\_width}}
\label{\detokenize{prog_desc:plot-width}}
\sphinxstylestrong{Description:} width of plot (in cm)

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} 24.0


\subsubsection{\sphinxstyleliteralintitle{-{-}plot\_yscale}}
\label{\detokenize{prog_desc:plot-yscale}}
\sphinxstylestrong{Description:} Y-axis min-max value, default is 1

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} 1.0


\subsubsection{\sphinxstyleliteralintitle{-{-}pre-check-only}}
\label{\detokenize{prog_desc:pre-check-only}}
\sphinxstylestrong{Description:} Only run the data pre-check (formely pre-dfoil.py)

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}pvalue}}
\label{\detokenize{prog_desc:pvalue}}\begin{description}
\item[{\sphinxstylestrong{Description:} minimum P-value cutoff for regions,}] \leavevmode
can specify one P-value for all four tests
or two separate ones for DFO/DIL and DFI/DOL
(or D1/D2 and D12 for ‘partitioned’)

\end{description}

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} {[}0.01, 0.01{]}


\subsubsection{\sphinxstyleliteralintitle{-{-}runlength}}
\label{\detokenize{prog_desc:runlength}}\begin{description}
\item[{\sphinxstylestrong{Description:} if two introgressing windows are separated}] \leavevmode
by this many windows of non-introgression
color in the intervening windows to
create a more continuous visual appearance

\end{description}

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 0


\subsubsection{\sphinxstyleliteralintitle{-{-}skip-pre-check}}
\label{\detokenize{prog_desc:skip-pre-check}}
\sphinxstylestrong{Description:} Skip running the data pre-check (formely pre-dfoil)

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}zerochar}}
\label{\detokenize{prog_desc:zerochar}}\begin{description}
\item[{\sphinxstylestrong{Description:} list of strings used in place of zeros}] \leavevmode
in the input file default is {[}“.”, “NA”{]}

\end{description}

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} {[}‘.’, ‘NA’{]}


\section{dfoil\_analyze}
\label{\detokenize{prog_desc:dfoil-analyze}}

\subsection{Description}
\label{\detokenize{prog_desc:id1}}
DFOIL: Directional introgression testing a five-taxon phylogeny
dfoil\_analyze: Given a dfoil output file, gives summary statistics to stdout
James B. Pease
\sphinxurl{http://www.github.com/jbpease/dfoil}


\subsection{Parameters}
\label{\detokenize{prog_desc:id2}}

\subsubsection{infile}
\label{\detokenize{prog_desc:infile}}
\sphinxstylestrong{Description:} dfoil output file

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id3}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}ndigits}}
\label{\detokenize{prog_desc:ndigits}}
\sphinxstylestrong{Description:} number of decimal places

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 3


\section{dfoil\_sim}
\label{\detokenize{prog_desc:dfoil-sim}}

\subsection{Description}
\label{\detokenize{prog_desc:id4}}
DFOIL: Directional introgression testing a five-taxon phylogeny
dfoil\_sim - simulation of sequences for testing dfoil
James B. Pease
\sphinxurl{http://www.github.com/jbpease/dfoil}


\subsection{Parameters}
\label{\detokenize{prog_desc:id5}}

\subsubsection{outputfile}
\label{\detokenize{prog_desc:outputfile}}
\sphinxstylestrong{Description:}  output site count filename

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id6}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}coaltimes}}
\label{\detokenize{prog_desc:coaltimes}}
\sphinxstylestrong{Description:} coalescent times in 4Ne units

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} (3, 2, 1, 1)


\subsubsection{\sphinxstyleliteralintitle{-{-}mdest}}
\label{\detokenize{prog_desc:mdest}}
\sphinxstylestrong{Description:} 1-based index of migration recipient population

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}mrate}}
\label{\detokenize{prog_desc:mrate}}\begin{description}
\item[{\sphinxstylestrong{Description:} per individual per generation migration}] \leavevmode
rate (default=5e-4)

\end{description}

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} 0.0005


\subsubsection{\sphinxstyleliteralintitle{-{-}msfile}}
\label{\detokenize{prog_desc:msfile}}\begin{description}
\item[{\sphinxstylestrong{Description:} use pre-computed ms output}] \leavevmode
file instead of running ms.

\end{description}

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}msource}}
\label{\detokenize{prog_desc:msource}}
\sphinxstylestrong{Description:} 1-based index of migration source population

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}mspath}}
\label{\detokenize{prog_desc:mspath}}
\sphinxstylestrong{Description:} path to ms executable

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} ms


\subsubsection{\sphinxstyleliteralintitle{-{-}mtimes}}
\label{\detokenize{prog_desc:mtimes}}
\sphinxstylestrong{Description:} time bounds for the migration period

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}mu}}
\label{\detokenize{prog_desc:mu}}\begin{description}
\item[{\sphinxstylestrong{Description:} per site per generation mutation rate}] \leavevmode
(default=7e-9)

\end{description}

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} 7e-09


\subsubsection{\sphinxstyleliteralintitle{-{-}nconverge}}
\label{\detokenize{prog_desc:nconverge}}
\sphinxstylestrong{Description:} number of convergent sites per window

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 0


\subsubsection{\sphinxstyleliteralintitle{-{-}nloci}}
\label{\detokenize{prog_desc:nloci}}
\sphinxstylestrong{Description:} number of windows to simulate

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100


\subsubsection{\sphinxstyleliteralintitle{-{-}popsize}}
\label{\detokenize{prog_desc:popsize}}
\sphinxstylestrong{Description:} Ne, effective population size (default=1e6)

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 1000000.0


\subsubsection{\sphinxstyleliteralintitle{-{-}quiet}}
\label{\detokenize{prog_desc:quiet}}
\sphinxstylestrong{Description:} suppress screen output

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}recomb}}
\label{\detokenize{prog_desc:recomb}}\begin{description}
\item[{\sphinxstylestrong{Description:} per site per generation recombination rate}] \leavevmode
(default=0)

\end{description}

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} 0.0


\subsubsection{\sphinxstyleliteralintitle{-{-}rho}}
\label{\detokenize{prog_desc:rho}}\begin{description}
\item[{\sphinxstylestrong{Description:} specific rho = 4*Ne*mu instead of using}] \leavevmode
\textendash{}recomb

\end{description}

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}window}}
\label{\detokenize{prog_desc:window}}
\sphinxstylestrong{Description:} length (bp) of windows

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100000


\section{fasta2dfoil}
\label{\detokenize{prog_desc:fasta2dfoil}}

\subsection{Description}
\label{\detokenize{prog_desc:id7}}
DFOIL: Directional introgression testing a five-taxon phylogeny
James B. Pease
\sphinxurl{http://www.github.com/jbpease/dfoil}

fasta2dfoil -
This script takes one or more FASTA files containing
5 or 4 taxa and counts site patterns for use in DFOIL/Dstat analysis.
To combine multiple FASTA files, each file should be sequences
from one locus (i.e., one entry in the final table) and
the names of sequences must be identical in all files.

Example Usage:

python3 fasta2dfoil.py INPUT.fasta \textendash{}out OUTPUT.fasta         \textendash{}names TAXA1,TAXA2,TAXA3,TAXA4


\subsection{Parameters}
\label{\detokenize{prog_desc:id8}}

\subsubsection{fastafile}
\label{\detokenize{prog_desc:fastafile}}\begin{description}
\item[{\sphinxstylestrong{Description:} one or more input fasta}] \leavevmode
files for each locus

\end{description}

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id9}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}names/-n} (required)}
\label{\detokenize{prog_desc:names-n-required}}\begin{description}
\item[{\sphinxstylestrong{Description:} Order of the 5 (or 4) taxa separated by commas.}] \leavevmode
Names must be  consistent in all input files,
outgroup should be last

\end{description}

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}out/-o} (required)}
\label{\detokenize{prog_desc:out-o-required}}
\sphinxstylestrong{Description:} output count file, one entry per fasta

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\chapter{Version History}
\label{\detokenize{releases:version-history}}\label{\detokenize{releases::doc}}

\section{2017-08-25}
\label{\detokenize{releases:id1}}
Updated precheck to include ‘dstat’ mode. Other fixes for precheck code.


\section{2017-06-14}
\label{\detokenize{releases:id2}}
Major upgrade to Sphinx documentation. Integrated the pre-check (formerly pre-dfoil.py) into the main script. Minor fixes to syntax.


\section{2017-01-29}
\label{\detokenize{releases:id3}}
Fixes for visual graph outputs, \sphinxtitleref{\textendash{}plot\_labels} has been fixed, replaced \sphinxtitleref{colornoanc} and \sphinxtitleref{colornoancdark} with \sphinxtitleref{\textendash{}plot\_noanc} option, changes to color pallette for code comprehension. Replaced \sphinxtitleref{\textendash{}plot-path} with just \sphinxtitleref{\textendash{}plot}, and \sphinxtitleref{\textendash{}plot show} is now deprecated.


\section{2015-11-23}
\label{\detokenize{releases:id4}}
More fixes for Python3 compability, added ‘pre-dfoil.py’ that checks for issues in count files before running dfoil.py


\section{2015-04-17}
\label{\detokenize{releases:id5}}
Minor updates and citation information, Publication Release Version


\section{2014-04-28}
\label{\detokenize{releases:id6}}
Fixes for Python3 compatibility


\section{2014-02-07}
\label{\detokenize{releases:id7}}
Re-release version on GitHub


\chapter{Indices and tables}
\label{\detokenize{index:indices-and-tables}}\begin{itemize}
\item {} 
\DUrole{xref,std,std-ref}{genindex}

\item {} 
\DUrole{xref,std,std-ref}{modindex}

\item {} 
\DUrole{xref,std,std-ref}{search}

\end{itemize}



\renewcommand{\indexname}{Index}
\printindex
\end{document}